tiled/access_policies.py

```diff
@@ -105,3 +127,502 @@ async def filters(self, node, principal, scopes, path_parts):
                     )
                 queries.append(query_filter(allowed))
         return queries
+
+
+def get_group_users(groupname):
+    """
+    A default method to retrieve group information
+    when compiling ACLs from tags. This should be
+    replaced by site-specific requirements.
+    """
+    import grp
+
+    return grp.getgrnam(groupname).gr_mem
+
+
+class InterningLoader(yaml.loader.BaseLoader):
+    pass
+
+
+def interning_constructor(loader, node):
+    value = loader.construct_scalar(node)
+    return intern(value)
+
+
+InterningLoader.add_constructor("tag:yaml.org,2002:str", interning_constructor)
+
+
+class LoadedTags(NamedTuple):
+    tags: dict
+    public: set
+    scopes: dict
+    owners: dict
+
+
+class TagBasedAccessPolicy:
+    def __init__(
+        self,
+        *,
+        provider,
+        tag_config,
+        group_parser,
+        scopes=None,
+    ):
+        self.provider = provider
+        self.tag_config_path = Path(tag_config)
+        self.scopes = scopes if (scopes is not None) else ALL_SCOPES
+        self.read_scopes = [intern("read:metadata"), intern("read:data")]
+        self.reverse_lookup_scopes = [intern("read:metadata"), intern("read:data")]
+        self.unremovable_scopes = [intern("read:metadata"), intern("write:metadata")]
+        self.admin_scopes = [intern("admin:apikeys")]
+        self.public_tag = intern("public").casefold()
```

Comment:
Should this intern the result of casefolding?
```suggestion
        self.public_tag = intern("public".casefold())
```

Reply:
Good point, this is technically true. I put the `casefold` there only to show intention for how this key is used, but yes this makes sense to do.